File: phases/18-language-paradigms-and-type-theory/11-oop-deep-inheritance-composition-mixins-traits/code/main.ts
interface Formatter {
  format(input: string): string;
}

class UppercaseFormatter implements Formatter {
  format(input: string): string {
    return input.toUpperCase();
  }
}

class Service {
  private readonly formatter: Formatter;

  constructor(formatter: Formatter) {
    this.formatter = formatter;
  }
  run(s: string): string {
    return this.formatter.format(s);
  }
}

const svc = new Service(new UppercaseFormatter());
console.log(svc.run("composed behavior"));

File: phases/18-language-paradigms-and-type-theory/11-oop-deep-inheritance-composition-mixins-traits/code/main.rs
trait Formatter {
    fn format(&self, s: &str) -> String;
}

struct Upper;

impl Formatter for Upper {
    fn format(&self, s: &str) -> String {
        s.to_uppercase()
    }
}

fn run_service(f: &dyn Formatter, input: &str) -> String {
    f.format(input)
}

fn main() {
    let u = Upper;
    println!("{}", run_service(&u, "trait composition"));
}
